### PR TITLE
feat(engine): feature flags

### PR DESCRIPTION
**Pull request by ekashida**
## Details

https://github.com/salesforce/lwc-rfcs/pull/3

Remaining tasks:
- [x] test utility for toggling feature flags
- [x] tests for rollup/babel plugins

```
export const ENABLE_FOO = true;
export const ENABLE_BAR = false;
export const ENABLE_BAZ = null;
```
Compiled in when `true`:
```
  if (ENABLE_FOO) {
    console.log('ENABLE_FOO');
  }
=>
  {
    console.log('ENABLE_FOO');
  }
```
The following is not supported (and it doesn't seem like we need to):
```
  if (isTrue(ENABLE_FOO)) {
    console.log('isTrue(ENABLE_FOO)');
  }
=>
  if (isTrue(ENABLE_FOO)) {
    console.log('isTrue(ENABLE_FOO)');
  }
```
Compiled out when `false`:
```
  if (ENABLE_BAR) {
    console.log('ENABLE_BAR');
  }
=>
  // nada
```
Runtime feature flags:
```
  if (ENABLE_BAZ) {
    console.log('ENABLE_BAZ');
  }
=>
  if (globalThis.LWC_config.features.ENABLE_BAZ) {
    console.log('ENABLE_BAZ');
  }
```
```
  if (ENABLE_BAZ === true) {
    console.log('ENABLE_BAZ === true');
  }
=>
  if ((globalThis.LWC_config.features.ENABLE_BAZ) === true) {
    console.log('ENABLE_BAZ === true');
  }
```
```
  if (isTrue(ENABLE_BAZ)) 
    console.log('isTrue(ENABLE_BAZ)');
  }
=>
  if (isTrue(globalThis.LWC_config.features.ENABLE_BAZ)) {
    console.log('isTrue(ENABLE_BAZ)');
  }
```

## Does this PR introduce breaking changes?

* ✅ `No, it does not introduce breaking changes.`

### CODE REVIEW

File: packages/@lwc/engine/scripts/rollup/rollup.config.js [hunk 1]
```diff
@@ -12,6 +12,9 @@ const { version } = require('../../package.json');
 const entry = path.resolve(__dirname, '../../src/framework/main.ts');
 const targetDirectory = path.resolve(__dirname, '../../dist/');
 
+const rollupLwcFeatures = require('@lwc/features/src/rollup-plugin');
```
Comment (ekashida) [inline packages/@lwc/engine/scripts/rollup/rollup.config.js]: I'm not so sure about how these plugins are accessed but I do think that they belong in the same package.

File: packages/@lwc/engine/scripts/rollup/rollup.config.js [hunk 2]
```diff
@@ -45,6 +48,7 @@ function rollupConfig({ format = 'es' } = {}) {
                 typescript,
                 include: TS_WHITELIST,
             }),
+            rollupLwcFeatures({ featureFlags }),
```
Comment (ekashida) [inline packages/@lwc/engine/scripts/rollup/rollup.config.js]: We could also make the module id (`@lwc/features`) and runtime member expression (`global.LWC_config.features`) configurable.

File: packages/@lwc/features/src/babel-plugin/index.js [hunk 3]
```diff
@@ -1,6 +1,10 @@
 const template = require('@babel/template');
 
-const buildRequire = template.expression('global.LWC_config.features.FEATURE_FLAG');
+const buildRequire = template.expression(`
+    global.LWC_config &&
```
Comment (caridy) [inline packages/@lwc/features/src/babel-plugin/index.js]: I think we can guarantee `global.LWC_config.features` to always be present by setting up that during initialization of the config code.

File: packages/@lwc/features/src/flags.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+type FeatureFlag = boolean | null;
```
Comment (caridy) [inline packages/@lwc/features/src/flags.ts]: in this file... at the very top, we should make sure that LWC_config exists... and it has a `features` record as well, even if it is just empty, this will guarantee that anyone importing from config is going to get that set up on front, not need to be defensive... the tree-shaking mechanism will preserve initializations.

File: packages/@lwc/features/package.json [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    "license": "MIT",
+    "scripts": {
+        "build": "tsc && tsc --project tsconfig.commonjs.json",
+        "clean": "rm -rf dist/"
+    },
+    "files": [
+        "dist/"
+    ],
+    "dependencies": {
+        "@babel/core": "~7.5.5",
```
Comment (ravijayaramappa) [inline packages/@lwc/features/package.json]: Why not use the @babel/core version from the root package.json?
@lwc/template-compiler also has @babel/template as a dependency, it can be lifted up to the root package.json

Comment (ekashida) [reply packages/@lwc/features/package.json]: I think it's ok to lift up devDeps to the root, but we should be explicit with dependencies.

Comment (caridy) [reply packages/@lwc/features/package.json]: yes, this should definitely be dev deps.

File: packages/@lwc/features/src/flags.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+type FeatureFlag = boolean | null;
+
+// Initialize the global configuration object if it isn't initialized already.
+const globalAny: any = global;
```
Comment (caridy) [inline packages/@lwc/features/src/flags.ts]: should we do `globalThis`? it will require a polyfill. How is that `global` (from node) transformed by our compile step for browsers? is it `windows`?

File: packages/@lwc/engine/scripts/rollup/rollup.config.js (hunk 1 above)
Comment (pmdartus) [reply packages/@lwc/engine/scripts/rollup/rollup.config.js]: A package should either expose a babel transformer or a rollup plugin. I don't think this package should expose both.

IMO, I would rather export a babel plugin instead and not a rollup config since a can use the babel transform via `babel-rollup-plugin`.

Comment (pmdartus) [inline packages/@lwc/engine/scripts/rollup/rollup.config.js]: Move the import about the relative file resolutions.

File: packages/@lwc/engine/scripts/rollup/rollup.config.js [hunk 7]
```diff
@@ -12,6 +12,9 @@ const { version } = require('../../package.json');
 const entry = path.resolve(__dirname, '../../src/framework/main.ts');
 const targetDirectory = path.resolve(__dirname, '../../dist/');
 
+const rollupLwcFeatures = require('@lwc/features/src/rollup-plugin');
+const featureFlags = require('@lwc/features');
```
Comment (pmdartus) [inline packages/@lwc/engine/scripts/rollup/rollup.config.js]: Couldn't the `featureFlags` parameter that you pass to the transformer be optional? This way you don't need to import the default feature flags to pass it again to the transformer.

File: packages/@lwc/engine/scripts/rollup/rollup.config.js (hunk 2 above)
Comment (pmdartus) [inline packages/@lwc/engine/scripts/rollup/rollup.config.js]: Why do we need to have 2 packages the distributed version of the engine don't have any reference to the original feature flags in the code?

File: packages/@lwc/features/package.json [hunk 8]
```diff
@@ -0,0 +1,19 @@
+{
+    "name": "@lwc/features",
```
Comment (pmdartus) [inline packages/@lwc/features/package.json]: In the current state of the PR, this package can't be used outside the monorepo. I think it would be made private for now.

File: packages/@lwc/features/src/rollup-plugin/index.js [hunk 9]
```diff
@@ -0,0 +1,19 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+const babel = require('@babel/core');
+const babelPlugin = require('../babel-plugin');
+
+module.exports = function rollupLwcFeatures(options = {}) {
```
Comment (pmdartus) [inline packages/@lwc/features/src/rollup-plugin/index.js]: I would rather use https://github.com/rollup/rollup-plugin-babel and use the transformer in this package than creating here a rollup plugin.

**Comment by diervo** [comment]
Wait for my review to merge this please will try to do it this week :)

File: packages/@lwc/engine/scripts/rollup/rollup.config.js (hunk 1 above)
Comment (ekashida) [reply packages/@lwc/engine/scripts/rollup/rollup.config.js]: @pmdartus `@lwc/engine` currently uses rollup so unless we want to refactor the current build, we should probably go with the rollup plugin. Also, it seems that `babel-rollup-plugin` is designed to generate a rollup bundle from files compiled by babel, which does not fit the use case for `@lwc/engine`.

Let me know if I misunderstood your feedback. If you prefer, I can move the rollup plugin wrapper into `@lwc/engine` since it's basically a one-liner.

File: packages/@lwc/engine/scripts/rollup/rollup.config.js (hunk 2 above)
Comment (ekashida) [reply packages/@lwc/engine/scripts/rollup/rollup.config.js]: I didn't understand this feedback, could you rephrase?

Comment (pmdartus) [reply packages/@lwc/engine/scripts/rollup/rollup.config.js]: Sorry.

I was wandering, who else except the engine would consume the feature flag package?

Comment (ekashida) [reply packages/@lwc/engine/scripts/rollup/rollup.config.js]: The compiler could eventually consume the package as well.

File: packages/@lwc/features/package.json (hunk 5 above)
Comment (ekashida) [reply packages/@lwc/features/package.json]: Was able to eliminate the dependency altogether by moving the rollup plugin wrapper into engine.

File: packages/@lwc/features/README.md [hunk 10]
```diff
@@ -0,0 +1,62 @@
+# Features
+
+[RFC](https://github.com/salesforce/lwc-rfcs/blob/master/text/0111-feature-flags.md)
+
+## Compile-time flags
+
+## Runtime flags
+
+Runtime flags are disabled by default and can be enabled or disabled via
+`globalThis.LWC_config.features`. Set the flag to `true` to enable and set the
```
Comment (caridy) [inline packages/@lwc/features/README.md]: This paragraph is a little bit confusing, who should set that there? and when? why is that affecting the compilation?

File: packages/@lwc/features/src/__tests__/flags.spec.ts [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+                console.log('isTrue(ENABLE_FEATURE_TRUE)');
+            }
+            if (!ENABLE_FEATURE_FALSE) {
+                console.log('!ENABLE_FEATURE_FALSE');
+            }
+        `,
+        output: `
+            import { ENABLE_FEATURE_TRUE, ENABLE_FEATURE_FALSE, ENABLE_FEATURE_NULL, runtimeFlags } from '@lwc/features';
+
+            if (ENABLE_FEATURE_NULL === null) {
```
Comment (caridy) [inline packages/@lwc/features/src/__tests__/flags.spec.ts]: I'm not sure why this is not transformed to `runtimeFlags.ENABLE_FEATURE_NULL`

File: packages/@lwc/features/src/__tests__/flags.spec.ts [hunk 12]
```diff
@@ -0,0 +1,307 @@
...
+              if (ENABLE_FEATURE_TRUE) {
+                console.log('ENABLE_FEATURE_TRUE');
+              }
+            }
+        `,
+    },
+    'should not transform feature flags when used with a ternary operator': {
+        code: `
+            import { ENABLE_FEATURE_NULL } from '@lwc/features';
+            console.log(ENABLE_FEATURE_NULL ? 'foo' : 'bar');
```
Comment (caridy) [inline packages/@lwc/features/src/__tests__/flags.spec.ts]: why not?

File: packages/@lwc/features/src/__tests__/flags.spec.ts [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+                if (ENABLE_FEATURE_TRUE) {
+                    console.log('this looks like a bad idea');
+                }
+            }
+        `,
+        output: `
+            import { ENABLE_FEATURE_TRUE, ENABLE_FEATURE_NULL, runtimeFlags } from '@lwc/features';
+
+            if (runtimeFlags.ENABLE_FEATURE_NULL) {
+              if (ENABLE_FEATURE_TRUE) {
```
Comment (caridy) [inline packages/@lwc/features/src/__tests__/flags.spec.ts]: why not?

File: packages/@lwc/features/src/__tests__/flags.spec.ts [hunk 14]
```diff
@@ -0,0 +1,307 @@
...
+                }
+
+                if (runtimeFlags.ENABLE_FEATURE_NULL) {
+                  console.log('ENABLE_FEATURE_NULL');
+                }
+            `,
+        },
+        'should transform runtime flag lookups into compile-time flags': {
+            code: `
+                if (runtimeFlags.ENABLE_FEATURE_TRUE) {
```
Comment (caridy) [inline packages/@lwc/features/src/__tests__/flags.spec.ts]: I don't understand this code... why is this using `runtimeFlags`?

File: packages/@lwc/features/src/__tests__/flags.spec.ts [hunk 15]
```diff
@@ -0,0 +1,307 @@
...
+            `,
+            output: `
+                console.log(runtimeFlags.ENABLE_FEATURE_NULL ? 'foo' : 'bar');
+            `,
+        },
+        'should not transform nested feature flags': {
+            code: `
+                import { ENABLE_FEATURE_TRUE, ENABLE_FEATURE_NULL } from '@lwc/features';
+                if (ENABLE_FEATURE_NULL) {
+                    if (ENABLE_FEATURE_TRUE) {
```
Comment (caridy) [inline packages/@lwc/features/src/__tests__/flags.spec.ts]: why not?

File: packages/@lwc/features/src/__tests__/flags.spec.ts (hunk 11 above)
Comment (ekashida) [reply packages/@lwc/features/src/__tests__/flags.spec.ts]: The transform is dead simple and only kicks in if the if-test is a simple identifier. Since the if-test is a binary expression, we ignore it.

File: packages/@lwc/features/src/__tests__/flags.spec.ts (hunk 12 above)
Comment (ekashida) [reply packages/@lwc/features/src/__tests__/flags.spec.ts]: I went with a simple implementation where feature flags only work for if-statements since I couldn't think of a scenario where that wouldn't be sufficient. I can transform feature flags without any restrictions if we think that's useful.

File: packages/@lwc/features/src/__tests__/flags.spec.ts (hunk 13 above)
Comment (ekashida) [reply packages/@lwc/features/src/__tests__/flags.spec.ts]: Nested feature flags sounds very complicated so I'm not sure we want to support that. We only need to remove a single line of code if we change our minds.

Comment (ekashida) [reply packages/@lwc/features/src/__tests__/flags.spec.ts]: To expand a little more, the number of possible outcomes is exponential when it comes to nesting. We should avoid that unless it's really needed.

File: packages/@lwc/features/src/__tests__/flags.spec.ts (hunk 14 above)
Comment (ekashida) [reply packages/@lwc/features/src/__tests__/flags.spec.ts]: This is to align with our build process where the dist files that are generated by rollup in `@lwc/engine` are processed again by rollup in a subsequent build for production.
